modxcms/revolution
# Add PSR-7/17/18 HTTP services

### What does it do?

Provides a standardized way of sending HTTP requests using Guzzle (or other compatible clients).

This should also be seen as deprecating modRest, alongside modRestClient which was already marked as such.

### Why is it needed?

Was working on fixing the package manager to no longer use modRestClient (#15469), however instead of trying to do it all ourselves we really ought to make use of PHP-wide standards. As it happens, PSR-7, PSR-17 and PSR-18 clearly outlines how to work with HTTP requests, so this implements that.

For the default implementation I've selected Guzzle. It's the de-facto standard, compliant with these specs, and actively supported matching our minimum requirements (PHP 7.2) as it happens.

### How to test

Here's an example GET request as it may appear in a snippet. Notice how it only uses interfaces, decoupling the implementation entirely from the calling code.

```php
<?php
$client = $modx->services->get(\Psr\Http\Client\ClientInterface::class);
$factory = $modx->services->get(\Psr\Http\Message\RequestFactoryInterface::class);
$request = $factory->createRequest('GET', 'http://ipinfo.io/')
    ->withHeader('Accept', 'application/json')
    ->withHeader('Content-Type', 'application/json');
try {
    $response = $client->sendRequest($request);
} catch (\Psr\Http\Client\ClientExceptionInterface $e) {
    $modx->log(1, $e->getMessage());
    return 'Error: ' . $e->getMessage();
}
$body = json_decode($response->getBody()->getContents());
return '<pre>' . $response->getStatusCode() . ' : ' . print_r($body, true) . '</pre>';
```

After pulling in the branch, it's necessary to run a `composer update` to load new dependencies. 

### Related issue(s)/PR(s)

Documentation: https://github.com/modxorg/Docs/pull/349/files
modRestClient deprecated: #15469

Following up on this assuming positive feedback/approval I will start implementing these into the package manager (which is the primary modRestClient implementation in core) which is a beta-blocking task.
